crates/c-api/src/ref.rs

```diff
@@ -1,24 +1,76 @@
+use crate::wasm_val_t;
+use std::any::Any;
 use std::os::raw::c_void;
-use wasmtime::ExternRef;
+use std::ptr;
+use wasmtime::{ExternRef, Func, Val};
 
-#[repr(C)]
+/// `*mut wasm_ref_t` is a reference type (`externref` or `funcref`), as seen by
+/// the C API. Because we do not have a uniform representation for `funcref`s
+/// and `externref`s, a `*mut wasm_ref_t` is morally a
+/// `Option<Box<Either<Option<ExternRef>, Option<Func>>>>`.
+///
+/// A null `*mut wasm_ref_t` is either a null `funcref` or a null `externref`
+/// depending on context (e.g. the table's element type that it is going into or
+/// coming out of).
+///
+/// Note: this is not `#[repr(C)]` because it is an opaque type in the header,
+/// and only ever referenced as `*mut wasm_ref_t`. This also lets us use a
+/// regular, non-`repr(C)` `enum` to define `WasmRefInner`.
 #[derive(Clone)]
 pub struct wasm_ref_t {
-    pub(crate) r: Option<ExternRef>,
+    pub(crate) r: WasmRefInner,
+}
+
+#[derive(Clone)]
+pub(crate) enum WasmRefInner {
+    ExternRef(Option<ExternRef>),
+    FuncRef(Option<Func>),
```

Comment:
It seems like there's two levels of `Option` in this representation, both here and in handling of `Option<Box<wasm_ref_t>>`. Is there a way that we could perhaps unify that so there's only one level at which null is introduced?